File: bin/create-concordance-pivot.R
#!/usr/bin/env Rscript
args = commandArgs(trailingOnly=TRUE)

# load necessary packages
library(tidyverse)
library(readxl)
library(writexl)

# parse command line arguments
cycle_table = args[1]
sample_manifest = args[2]

# create a vector of detailed report files
results_files <- list.files(path = ".", pattern = "_Detailed_Results")

# find available cycles
available_cycles <- c()
for (i in results_files){
  cycle <- unlist(strsplit(i, split = "\\."))[1] %>%
    str_sub(-1,-1) %>%
    as.integer()
  available_cycles <- c(available_cycles, cycle)
}

# read in optimal cycle table
cycle_table <- read.csv(cycle_table)

# remove unknown samples and correct column types
# cycle_table <- cycle_table[cycle_table$cycle!="U" & cycle_table$cycle!="X" &
#                              !is.na(cycle_table$cycle),]
# cycle_table$cycle <- as.numeric(cycle_table$cycle)

# Read in the list of SNPs
SNPs <- read_excel(results_files[1], sheet = 1, skip = 15)[1] %>%
  separate(ID, into = c("sample", "SNP"), sep = "-") %>%
  subset(select = "SNP")

# make an empty data frame that will subsequently be filled for each SNP
concordance_table <- data.frame("SNP" = sort(unique(SNPs$SNP)),	
                                "Assay" = NA,
                                "Cycle" = NA, 
                                "Comments" = NA)
# determine optimal cycles
cycle_table <- cycle_table[order(cycle_table$SNP),]
concordance_table$Cycle <- as.integer(cycle_table[cycle_table$SNP==concordance_table$SNP,"cycle"])
concordance_table <- concordance_table[concordance_table$Cycle!=0,]

# Create an empty list to store the excel dataframes
df_list = list()

# Read in the Excel files and store them in the list
for (file in results_files){
  i <- as.integer(strsplit(strsplit(file, split = "\\.")[[1]][1], split = "_cycle")[[1]][2])
  df_list[[i]] <- read_excel(file, skip = 15, sheet = 1)
}

# loop through each SNP to collate calls
for (snp in concordance_table$SNP){
  
  # determine optimal cycle
  optimal_cycle <- as.integer(cycle_table[cycle_table$SNP==snp,"cycle"])
  
  if (!(optimal_cycle %in% available_cycles)){
    break
  }
  
  # read in the correct BioMark output file based on that cycle
  data <-  df_list[[optimal_cycle]]
  
  # split sample IDs and SNP IDs into their own columns
  data <- separate(data, ID, into = c("sample", "SNP"), sep = "-")
  
  snp_table <- data[data$SNP==snp,]
  new_index <- which(concordance_table$SNP==snp)
  
  concordance_table[new_index, "Assay"] <- unique(snp_table$Assay)
  concordance_table[new_index, "Cycle"] <- optimal_cycle
  
  if (snp == concordance_table$SNP[1]) {

    for (animal in sort(unique(snp_table$Name))){

      call <- snp_table[snp_table$Name==animal,"Final"][1,] %>%
        as.character()
      concordance_table <- cbind(concordance_table, as.data.frame(NA))
      concordance_table[new_index,ncol(concordance_table)] <- call
      colnames(concordance_table)[ncol(concordance_table)] <- animal

    }

  } else if (ncol(concordance_table[,5:ncol(concordance_table)]) != length(unique(data$sample))+1) {
    
    for (animal in sort(unique(snp_table$Name))){
      
      call <- snp_table[snp_table$Name==animal,"Final"][1,] %>%
        as.character()
      concordance_table[new_index,animal] <- call
      
    }
    
  } else {
    break
  }
  
}

# change assay names to SNP names
concordance_table$SNP <- str_replace(concordance_table$SNP, "A", "SNP")

# read in and format the sample manifest
samples <- read_excel(sample_manifest, skip = 1)
samples <- samples[!is.na(samples$`Animal ID`) & !is.na(samples$Relationship),]

# parse out the relationship "groups"
samples <- subset(samples, select = c("Animal ID", "Relationship"))
samples$group <- str_remove(samples$Relationship, "Dam ") %>%
  str_remove("Progeny ") %>% 
  as.numeric()

# go through each group and add them to new vectors that will eventually become
# new column headers for the concordance pivot
ordered_samples <- c()
ordered_relations <- c()
for (i in unique(samples$group)){
  
  # go group by group
  sub <- samples[samples$group==i,]
  
  # first add the dams
  ordered_samples <- c(ordered_samples, as.character(sub[grepl("Dam", sub$Relationship), "Animal ID"]))
  ordered_relations <- c(ordered_relations, as.character(sub[grepl("Dam", sub$Relationship), "Relationship"]))
  
  # then add the progeny
  ordered_samples <- c(ordered_samples, as.character(sub[grepl("Progeny", sub$Relationship), "Animal ID"]))
  ordered_relations <- c(ordered_relations, as.character(sub[grepl("Progeny", sub$Relationship), "Relationship"]))
  
}

# bind together the vectors as a data frame
# ordered_samples <- as.data.frame(ordered_samples)
# ordered_relations <- as.data.frame(ordered_relations)
# new_order <- t(cbind(ordered_samples,ordered_relations))

# define where the animals start in the concordance pivot, and sort them according
# to the new sample vector
animal_start <- which(colnames(concordance_table)=="Comments")+1
animal_cols <- concordance_table[,animal_start:ncol(concordance_table)]
concordance_table <- concordance_table[,1:(animal_start-1)]
for (i in 1:length(ordered_samples)){
  
  animal <- ordered_samples[i]
  relation <- ordered_relations[i]
  sub <- c(animal, relation, animal_cols[,animal])
  
  if (i == 1){
    
    new_cols <- data.frame(sub)
    
  } else {
    
    new_cols <- cbind(new_cols, sub)
    
  }
  
}

# prepare new rows for the concordance table
new_row1 <- c("Animal ID", NA, NA, NA)
new_row2 <- colnames(concordance_table)
colnames(concordance_table) <- NULL
concordance_table <- rbind(new_row1, new_row2, concordance_table)
concordance_table <- cbind(concordance_table, new_cols)
colnames(concordance_table) <- NULL

# write output file for viewing in excel
write_xlsx(concordance_table, "concordance-table.xlsx", col_names = F)

File: bin/create-concordance-pivot.py
#!/usr/bin/env python3

import sys
import os
import polars as pl
import xlsxwriter

# Check if the input TSV file is a symlink, and if it is, follow the symlink
cycle_table = os.path.realpath(sys.argv[1])
sample_manifest = os.path.realpath(sys.argv[2])

# Find available cycles
results_files = [file for file in os.listdir('.') if "_Detailed_Results" in file]
available_cycles = []
for i in results_files:
    cycle = int(i.split('.')[0].split("cycle")[-1])
    available_cycles.append(cycle)

# Read in optimal cycle table and sort by SNP name
cycle_table = pl.read_csv(cycle_table).sort('SNP')

# Read in the list of SNPs
SNPs = pl.read_excel(results_files[0], read_csv_options={"skip_rows": 15})
SNPs = SNPs.select(pl.col("ID").str.split(by="-")).with_columns(
        pl.struct(
            [
                pl.col('ID').arr.get(i).alias(
                    'Sample' if i==0 else 'SNP') 
                for i in range(2)
            ]
        ).alias('id_struct')
        )
SNPs = SNPs.unnest('id_struct')
SNPs = SNPs.drop('ID')

# Create an empty data frame that will subsequently be filled for each SNP
concordance_table = pl.DataFrame({
    'SNP': SNPs['SNP'].unique().sort(),
    'Assay': "None",
    'Cycle': cycle_table['cycle'],
    'Comments': None
})

# Filter out SNPs with not optimal cycle
concordance_table = concordance_table.filter(pl.col('Cycle') != 0)

# Loop through each SNP to collate calls
for i in range(len(concordance_table)):
  
    # define SNP for this iteration
    snp = concordance_table['SNP'][i]
  
    # Determine optimal cycle
    optimal_cycle = cycle_table.filter(pl.col('SNP') == snp)['cycle'].item()
    
    if optimal_cycle not in available_cycles:
        break

    # Read in the correct BioMark output file based on that cycle
    data = pl.read_excel(f"28637_BMX003_Detailed_Results_cycle{optimal_cycle}.xlsx", read_csv_options={"skip_rows": 15})
    snp_sample = data.select(pl.col("ID").str.split(by="-")).with_columns(
        pl.struct(
            [
                pl.col('ID').arr.get(i).alias(
                    'Sample' if i==0 else 'SNP') 
                for i in range(2)
            ]
        ).alias('id_struct')
        ).unnest('id_struct').drop('ID')
    data = data.with_columns([
      snp_sample['SNP'].alias('SNP'),
      snp_sample['Sample'].alias('Sample')])

    # make table of SNPs
    snp_table = data.filter(pl.col('SNP') == snp)
    
    # Record SNP assay code in the concordance pivot
    concordance_table = concordance_table.with_columns([
      concordance_table['Assay'].set_at_idx(i,snp_table['Assay'].unique().item()).alias('Assay')
    ])

    if snp == concordance_table['SNP'][0]:
        for animal in sorted(snp_table['Name'].unique()):
            call = snp_table.filter(pl.col('Name') == animal)['Final'].unique()[0]
            concordance_table = concordance_table.with_columns([
              pl.col('Assay').alias(animal)
            ])
            concordance_table = concordance_table.with_columns([
              concordance_table[animal].set_at_idx(i,call).alias(animal)
            ])
    elif len(concordance_table[:, 5:]) != len(data['Sample'].unique()) + 1:
        for animal in sorted(snp_table['Name'].unique()):
            call = snp_table.filter(pl.col('Name') == animal)['Final'].unique()[0]
            concordance_table = concordance_table.with_columns([
              concordance_table[animal].set_at_idx(i,call).alias(animal)
            ])
    else:
        break

# Change assay names to SNP names
concordance_table = concordance_table.with_columns(concordance_table['SNP'].str.replace(r"A", "SNP").alias('SNP'))

# Read in and format the sample manifest
samples = pl.read_excel(sample_manifest, read_csv_options={"skip_rows": 1}).drop_nulls(subset=['Animal ID', 'Relationship'])

# Parse out the relationship "groups"
samples = samples.select('Animal ID', 'Relationship').with_columns(
    samples['Relationship'].str.replace('Dam ', '').str.replace('Progeny ', '').cast(pl.Int64, strict=False).alias('group')
)

# Go through each group and add them to new vectors that will eventually become 
# new column headers for the concordance pivot
ordered_samples = []
ordered_relations = []
for i in samples['group'].unique():
    sub = samples.filter(pl.col('group') == i)
    ordered_samples.append(sub.filter(
      pl.col('Relationship').str.contains('Dam')
    ).select('Animal ID').item())
    ordered_samples.append(sub.filter(
      pl.col('Relationship').str.contains('Progeny')
    ).select('Animal ID').item())
    ordered_relations.append(sub.filter(
      pl.col('Relationship').str.contains('Dam')
    ).select('Relationship').item())
    ordered_relations.append(sub.filter(
      pl.col('Relationship').str.contains('Progeny')
    ).select('Relationship').item())

# Bind together the vectors as a data frame
animal_cols = concordance_table[:,4:]
for i in range(len(ordered_samples)):
  
  animal = ordered_samples[i]
  relation = ordered_relations[i]
  sub = pl.Series([animal, relation])
  sub = sub.append(animal_cols[animal], append_chunks=False)
  
  if i == 0:
    new_cols = pl.DataFrame({"new_col_0": sub})
  else:
    col = "new_col_"
    new_cols = new_cols.with_columns([
      sub.alias("{}{}".format(col, i))
    ])

# Prepare new rows for the concordance table
new_row1 = pl.DataFrame(["Animal Name", None, None, None]).transpose(include_header=False)
new_row2 = pl.DataFrame(concordance_table[:, :4].columns).transpose(include_header=False)

# combing new rows 1 and 2 and the non-animal-column portion of the concordance
# table with a row-bind
final_table = new_row1.vstack(new_row2)
non_call_rows = concordance_table[:,:4]
for i in range(len(non_call_rows.columns)):
  old_col = non_call_rows.columns[i]
  col = "column_"
  non_call_rows = non_call_rows.rename({old_col: "{}{}".format(col, i)})
  non_call_rows = non_call_rows.with_columns([
    non_call_rows[:,i].cast(pl.Utf8).alias("{}{}".format(col, i))
    ])
final_table = final_table.vstack(non_call_rows)

# tack on the animal-column portion of the concordance table with a column-bind
final_table = final_table.hstack(new_cols)

# Write output file for viewing in Excel
final_table.write_excel('concordance-table.xlsx', autofit=True, has_header=False)
